config/file.go

```diff
@@ -0,0 +1,323 @@
+	fs.configLock.Lock()
+	var unlockOnce sync.Once
+	defer func() {
+		unlockOnce.Do(func() {
+			fs.configLock.Unlock()
+		})
+	}()
+
+	oldCfg := fs.config
+
+	// TODO: disallow attempting to save a directly modified config (comparing pointers). This
+	// wouldn't be an exhaustive check, given the use of pointers throughout the data
+	// structure, but might prevent common mistakes. Requires upstream changes first.
+	// if newCfg == oldCfg {
+	// 	return nil, errors.New("old configuration modified instead of cloning")
+	// }
+
+	newCfg = newCfg.Clone()
+	newCfg.SetDefaults()
+
+	// Sometimes the config is received with "fake" data in sensitive fields. Apply the real
+	// data from the existing config as necessary.
+	desanitize(oldCfg, newCfg)
+
+	if err := newCfg.IsValid(); err != nil {
+		return nil, errors.Wrap(err, "new configuration is invalid")
+	}
+
+	if *oldCfg.ClusterSettings.Enable && *oldCfg.ClusterSettings.ReadOnlyConfig {
+		return nil, ReadOnlyConfigurationError
+	}
+
+	fs.needsSave = true
+
+	// Ideally, Set would persist automatically and abstract this completely away from the
+	// client. Doing so requires a few upstream changes first, so for now an explicit Save()
+	// remains required.
+	// if err := fs.persist(newCfg); err != nil {
+	// 	return nil, errors.Wrap(err, "failed to persist")
+	// }
+
+	fs.config = newCfg
+
+	unlockOnce.Do(func() {
+		fs.configLock.Unlock()
+	})
+
+	// Notify listeners synchronously. Ideally, this would be asynchronous, but existing code
+	// assumes this and there would be increased complexity to avoid racing updates.
+	fs.invokeConfigListeners(oldCfg, newCfg)
+
+	return oldCfg, nil
+}
+
+// Patch merges the given config with the current configuration, without updating the backing store.
+func (fs *fileStore) Patch(*model.Config) (*model.Config, error) {
+	fs.configLock.Lock()
+	defer fs.configLock.Unlock()
+
+	panic("patch is not yet implemented")
+
+	return fs.config, nil
+}
+
+// serialize converts the given configuration into JSON bytes for persistence.
+func (fs *fileStore) serialize(cfg *model.Config) ([]byte, error) {
+	return json.MarshalIndent(cfg, "", "    ")
+}
+
+// persist writes the configuration to the configured file.
+func (fs *fileStore) persist(cfg *model.Config) error {
+	fs.stopWatcher()
+
+	b, err := fs.serialize(cfg)
+	if err != nil {
+		return errors.Wrap(err, "failed to serialize")
+	}
+
+	err = ioutil.WriteFile(fs.path, b, 0644)
+	if err != nil {
+		return errors.Wrap(err, "failed to write file")
+	}
+
+	fs.needsSave = false
+	if fs.watch {
+		if err = fs.startWatcher(); err != nil {
+			mlog.Error("failed to start config watcher", mlog.String("path", fs.path), mlog.Err(err))
+		}
+	}
+
+	return nil
+}
+
+// Load updates the current configuration from the backing store.
+func (fs *fileStore) Load() (needsSave bool, err error) {
+	var f io.ReadCloser
+
+	f, err = os.Open(fs.path)
+	if os.IsNotExist(err) {
```

Comment:
It's perfectly fine to start out without a file: the default behaviour is to initialize with the defaults. Now, we no longer require `default.json` to be located, and instead just write out the defaults as codified in `model.Config`.